trin-core/src/portalnet/overlay_service.rs

```diff
@@ -1924,4 +2204,190 @@ mod tests {
+    }
+
+    #[tokio::test]
+    async fn test_advance_findnodes_query() {
+        let mut service = task::spawn(build_service());
+
+        let (_, bootnode) = generate_random_remote_enr();
+        let bootnodes = vec![bootnode.clone()];
+        let bootnode_node_id = bootnode.node_id();
+
+        let (_, target_enr) = generate_random_remote_enr();
+        let target_node_id = target_enr.node_id();
+
+        service.init_find_nodes_query_with_initial_enrs(&target_node_id, bootnodes);
+
+        // Test that the first query event contains a proper query ID and request to the bootnode
+        let event =
+            OverlayService::<IdentityContentKey, XorMetric, MockValidator>::query_event_poll(
+                &mut service.queries,
+            )
+            .await;
+        match event {
+            FindNodeQueryEvent::Waiting(query_id, node_id, request) => {
+                match request {
+                    Request::FindNodes(find_nodes) => {
+                        assert_eq!(
+                            find_nodes.distances.len(),
+                            service.findnodes_query_distances_per_peer
+                        );
+                    }
+                    _ => panic!(),
+                }
+                assert_eq!(query_id, QueryId(0));
+                assert_eq!(node_id, bootnode_node_id);
+            }
+            _ => panic!(),
+        }
+
+        // Create two ENRs to use as dummy responses to the query from the bootnode.
+        let (_, enr1) = generate_random_remote_enr();
+        let (_, enr2) = generate_random_remote_enr();
+        let node_id_1 = enr1.node_id();
+        let node_id_2 = enr2.node_id();
+
+        service.advance_query(bootnode, vec![enr1, enr2], QueryId(0));
+
+        let event =
+            OverlayService::<IdentityContentKey, XorMetric, MockValidator>::query_event_poll(
+                &mut service.queries,
+            )
+            .await;
+
+        // Check that the request is being sent to either node 1 or node 2. Keep track of which.
+        let first_node_id: Option<NodeId>;
+        match event {
+            FindNodeQueryEvent::Waiting(_, node_id, _) => {
+                assert!((node_id == node_id_1) || (node_id == node_id_2));
+                first_node_id = Some(node_id);
+            }
+            _ => panic!(),
+        }
+
+        let event =
+            OverlayService::<IdentityContentKey, XorMetric, MockValidator>::query_event_poll(
+                &mut service.queries,
+            )
+            .await;
+
+        // Check that a request is being sent to the other node.
+        let second_node_id = if first_node_id.unwrap() == node_id_1 {
+            node_id_2
+        } else {
+            node_id_1
+        };
+        match event {
+            FindNodeQueryEvent::Waiting(_, node_id, _) => {
+                assert_eq!(node_id, second_node_id);
+            }
+            _ => panic!(),
+        };
```

Comment:
We do not test that we get the expected result from a finished or timed out query here.

Reply:
Good point, I'll add that.